lib/puppet/type/agent_upgrade.rb

```diff
@@ -0,0 +1,30 @@
+Puppet::Type.newtype(:agent_upgrade) do
+  ensurable
+
+  newparam(:name, :namevar => true)
+
+  newparam(:source) do
+    newvalues /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+    validate do |value|
+      if value !~ /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+        fail "Please provide a valid http(s), puppet or unc path"
+      end
+      if value =~ /^https?/ && value !~ URI::regexp
+        fail("Please provide a valid URI")
```

Comment:
I'm actually worried about this one - it's not agnostic and the check should be implemented up in the public class, not in here. The check is going to need to be done anyway, implementing it here just means it needs to be reimplemented in the class. However I'm worried both about duplication of work and that one will be more restrictive than another, leading to incorrect documentation as such.

Plus it should likely be more agnostic.